examples/coarse_alignment.py

```diff
@@ -86,6 +89,46 @@
     current_shift += shift
     coarse_shifts[i + 1] = current_shift
 
+# optimize tilt axis angle
+grid_resolution = 1
+tilt_axis_grid = CubicBSplineGrid1d(resolution=grid_resolution, n_channels=1)
+tilt_axis_grid.data = torch.tensor([TILT_AXIS_ANGLE_PRIOR, ] * grid_resolution, dtype=torch.float32)
+interpolation_points = torch.linspace(0, 1, len(tilt_series))
+
+common_lines_optimiser = torch.optim.Adam(
+    tilt_axis_grid.parameters(),
+    lr=1,
+)
+
+for epoch in range(100):
+    # interpolate the grid
+    tilt_axis_angles = tilt_axis_grid(interpolation_points)
+
+    # for common lines each 2d image is projected perpendicular to the tilt axis, thus add 90 degrees
+    R = Rz(tilt_axis_angles + 90, zyx=False)[:, :2, :2]
+
+    projections = []
+    for i in range(len(tilt_series)):
+        p = project_real_2d(tilt_series[i] * coarse_alignment_mask, R[i:i+1])
+        projections.append((p - p.mean()) / p.std())
+
+    common_lines_optimiser.zero_grad()
+    loss = 0
+    for x, y in combinations(projections, 2):
+        loss = loss - (x * y).sum() / y.numel()
+    loss.backward()
+    common_lines_optimiser.step()
+
+    if not (epoch % 10):
+        print(epoch, loss.item(), tilt_axis_grid.data.mean())
+
+tilt_axis_prediction = tilt_axis_grid(interpolation_points).clone().detach()
+print('final tilt axis angle:', torch.unique(tilt_axis_prediction))
+
+# apply the shifts only after optimizing the axis angle, optimizing
+# the shifts on the aligned tilt-series seems to fail
+coarse_aligned_tilt_series = shift_2d(tilt_series, shifts=-coarse_shifts)
```

Comment:
huh, do you agree this feels unexpected? Do you think it's a wraparound thing?

Reply:
Absolutely and I couldn't let it rest, so managed to come up with a fix (see message below).